Event building update

mnieslony opened the pull request:
The Event Building has been extended in multiple aspects:
* Beam status information was added to the event matching and is now stored in the ANNIEEvent store. The `BeamDecoder` tool was added to incorporate this information during the EventBuilding process.
*  The hit finding process was incorporated into the EventBuilding chain and does not need to be run on the processed files anymore. This reduces the processed file size by a factor of ~100.
* Trigger information about extended readout windows is stored in the ANNIEEvent store now
* It is now possible to build events with 1 VME waveform missing (sometimes the RWM waveform seems to be missing)
* Additional event building options have been introduced, such as Tank+CTC and MRD+CTC (in addition to the standard Tank+MRD+CTC)
* More information is saved in the OrphanStore that can help debug causes for orphaned events.
* The `PreProcessTrigOverlap` toolchain was added. It should be run before the main EventBuilding process and provides trigger information from the previous and subsequent file that are both needed to match all available MRD+tank timestamps in a single file (the trigger data information overlaps with the previous and subsequent file)
* Timestamp matching cuts have been relaxed from 100ns to 300ns for tank/CTC timestamp matching. Sometimes the timestamps seem to be 256ns apart for no apparent reason (this is the minority of events, but we still want to have them in the data)
* Triggerword matching was extended to triggerword 31 (LED_Start) in addition to triggerword 33 (LED_ADC_Trigger). The LED_Start triggerword timestamp seems to match the tank timestamps much better in beam runs than the LED_ADC_Trigger timestamp.

DataModel changes:
* Two additional members were added to the `TriggerClass`: The Triggerword and the ExtendedWindow variables
* Two additional labels were added to the `MinibufferLabel` class, one for MRDCR triggers and one for AmBe events.

Other tool changes:
* The `DataSummary` tool was extended with more debugging information about orphans and built events
* The `LoadANNIEEvent` tool was fixed so it can read in normal ANNIEEvent BoostStores again (bugfix). A global event counter option was added in case one reads out multiple subsequent files from the same run and needs a global event numbering scheme.
* The `TriggerDataDecoder` tool was fixed so it can store multiple triggerwords per timestamp (bugfix)
* The `LoadRunInfo` tool was added and can be used to load information about the runs from a txt-file (contents for the txt-file can be copied from the SQL database page)
* The `PrintANNIEEvent` tool was extended to give a better overview of ANNIEEvent data files.

mnieslony committed:
Added new pairings in event building (tank/ctc, mrd/ctc)

mnieslony committed:
added automatic Run/SubRun/Part numbering scheme for processed files

mnieslony committed:
added Store/PlotDecodedTimestamp tools

mnieslony committed:
Put more information into the Orphan Event BoostStore

mnieslony committed:
Added EventBuilding support for 133 waveforms in case the RWM buffer not saved

mnieslony committed:
Added PreProcessTrigOverlap

mnieslony committed:
Update README of PreProcessTrigOverlap

mnieslony committed:
Update README of PlotDecodedTimestamps

mnieslony committed:
Update README of StoreDecodedTimestamps

mnieslony committed:
Fixed slowest timestream also for Tank+CTC event building

mnieslony committed:
Merge branch 'EventBuilding' of https://github.com/mnieslony/ToolAnalysis into EventBuilding

mnieslony committed:
Update README of DataSummary toolchain

mnieslony committed:
Update README of DataSummary tool

mnieslony committed:
Update README of DataDecoderTankCTC

mnieslony committed:
Update README of StoreDecodedTimestamps tool

mnieslony committed:
Update README of PlotDecodedTimestamps tool

mnieslony committed:
add MaxPEPlots tool

mnieslony committed:
Merge branch 'EventBuilding' of https://github.com/mnieslony/ToolAnalysis into EventBuilding

mnieslony committed:
Added bugfix for TriggerDataDecoder

mnieslony committed:
first version of HitBuilding in EventBuilding

mnieslony committed:
Added BeamStatus object to EventBuilding

mnieslony committed:
Added RawAcqSize object and TriggerData object to ANNIEEvent

mnieslony committed:
Added PrintANNIEEvent and BeamFetcher toolchains

mnieslony committed:
Update README of BeamDecoder

mnieslony committed:
Update README of BeamFetcher tool

mnieslony committed:
Adapted configfiles for other DataDecoder configurations

mnieslony committed:
Merge branch 'EventBuilding' of https://github.com/mnieslony/ToolAnalysis into EventBuilding

mnieslony committed:
Added LoadrunInfo tool

mnieslony committed:
Update README of LoadRunInfo

mnieslony committed:
Update README of BeamFetcher toolchain

mnieslony committed:
Delete README of DataDecoderMRDCTC toolchain

mnieslony committed:
Update README of PrintANNIEEvent toolchain

mnieslony committed:
Adjusted settings for PhaseIIADCCalibrator & Hit Fider

mnieslony committed:
Merge branch 'EventBuilding' of https://github.com/mnieslony/ToolAnalysis into EventBuilding

mnieslony committed:
Added additional MaxPEPlots features

mnieslony committed:
Updated DataSummary

mnieslony committed:
moved auxiliary tools to different branch

mnieslony committed:
Added functionality to detect potential dead PMTs in event building

mnieslony committed:
fix for small bug in PhaseIIADCCalibrator

mnieslony committed:
reversed order of ANNIEEventBuilder and PhaseIIADCCalibrator/HitFinder

marc1uk commented:
General remark: i know it's been said, but 111 files changed in one PR is a heckova lot.

To start with, many of the files changed come from adding new toolchains. :+1: Happy to have them, but they can very easily be put into one PR per toolchain, since they're generally independent of any other changes. Easy to review and approve.

Next separating all changes related to introducing a new or modified feature would be even better - e.g. 'Add support for multiple TriggerWords per event', 'add support for handling overlap between part files' etc. It makes it more clear on the scope of the changes, which makes checking for bugs and consistency much easier, and it means other developers who want to 'get up to speed' on a new feature can review the PR to see what it involves.

The number of commits here suggests these changes didn't all happen at once, so it shouldn't be too much work to split things up in this way. Just make a bunch of different branches along your git history (commit A->B->C [branch_1]->D->E->F[branch_2]->...), and put in PRs for each of those branches. If you specify the order in the comments ('PR ###, depends on PR ###'), we can make sure they get merged in the right order. Then github should automatically update the reference branch and set of changes associated to the next PR as each gets merged in, so that each new PR shows only the new changes and should be much easier to review.

OK some specifics:
https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-0f05b8999c4d7a65e71b207be61acd8115305eb7203c62acb0d121fab73e956dR118
this seems redundant (getting a variable then setting it to a constant).

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-0f05b8999c4d7a65e71b207be61acd8115305eb7203c62acb0d121fab73e956dR218
if these aren't needed, can we remove them to clean up the code?

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-cd39c98bd3b647f9384a07dfb2fb109d2be6d745b6566777e4a1d6ffccec7d6dR6
this seems handy, might be worth putting in ANNIEAlgorithms or something.

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-0f05b8999c4d7a65e71b207be61acd8115305eb7203c62acb0d121fab73e956dR338
whats the justification and implications of the 'MRD always being faster'?

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-cd39c98bd3b647f9384a07dfb2fb109d2be6d745b6566777e4a1d6ffccec7d6dR59
if this is part of the usual event building toolchain it'd be better to replace std::cout with Log with a lower verbosity for frequent printouts. Otherwise a really nicely written tool.

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-1efd41d9f010db9a4e71289b7f6e3f25066cf2de2e41499db687c3ee6ee2bd1aR41
Personally i'm not sure falling back to a default when given an unrecognised option value is a good idea. If **no** option is given, falling back to default is ok, but if an invalid option is given then maybe just bailing the toolchain is better. These warnings can very easily get lost in the wall of output, resulting in data being processed incorrectly. If the user asked for something, either give them what they asked for, or ask again. These are the kinds of situations i'd like to be able to force a toolchain to stop, with a finer grained control than a global setting for 'how to handle tool Initialize/Execute/Finalise returning false'. I guess there's always 'assert(false)', but that feels a bit crude... :slightly_frowning_face: 

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-1efd41d9f010db9a4e71289b7f6e3f25066cf2de2e41499db687c3ee6ee2bd1aR153
std::cout -> Log. May also be worth mentioning in the readme that the output filename will automatically be constructed from the Run number in most Timestamp modes.

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-def45d71da0c479e49aa2a92f396214f35f198409ef8937d76fc3d87dc5b1768R21
is this something we need to commit? Is this a default toolchain format, or just a debug setting?

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-b20cfa846b04df79ff29926087aa28e3e0df67adefd025c384fad5ad1135137eR2
https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-1fa0402ef729dada0e326e28a4e7de26d3358e5f9f9009b464b78ff809f0bd92R2
Personally i'm not a fan of toolchains cross-linking their config files. If it's just file with tool configuration variables, make a copy (the text files are small enough). For something like the geometry csv, it makes more sense to me for this not to be associated to a particular toolchain, and actually place it in something like the repository parent directory, the DataModel folder, or (probably the best option) create a dedicated 'configurations' folder with such global configuration files.

Perhaps we could even consider having a directory of 'default config files' for some tools. Currently we have many toolchains with things like the 'ADCHitFinderConfig', but with no real need to do anything non-standard. They're all just copy-pasting the same configuration file, which results in a lot of duplication and a lot of toolchains becoming outdated if the default configuration changes.

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-05a56d4173fd57c8f9841f5f62599f2610f4cf61eb0844edeb00a8265ebc81bfR1
Do we need separate toolchains for decoding TandAndMRD, TankAndMRDAndCTC, MRDAndCTC, TankAndCTC, TankAndCTCButNotMRDExceptThoseEventsWithAhhhhSoManyToolchains...?

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-f8a9562384e11f68dd4d5b44e66d64f6d900d87ef22ecd7e62a128473885c572R1
dummy files can be removed

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-2876438e4c810ff12d516d8f26659eaf7cea38e80b5658dc8fec0a0281a655eaR48
Please turn on whitespace visibility. :upside_down_face: There's a few others in this PR too (although I leave it to you to go through the 111 files to find them. ;) )

https://github.com/ANNIEsoft/ToolAnalysis/pull/187/files#diff-5cbbc194a79e9d787108be01e4964f066530ee70c0d8ef1b99d5d335727c9a04R68
Seems like the 'MCFlag' is a bit redundant with the 'is_mc' flag. I would personally prefer some automatic detection, which seems like it ought to be easy enough?

General comment: the event building process is getting increasingly convoluted, and we're getting more and more tools running into thousands of lines of code, with multiple "modes" of operation. e.g. the large fraction of code in the PhaseIIADCCalibrator and PhaseIIADCHitFinder tools within `if(mode_a){ ... } else { ... }` blocks.
It might be worth considering either splitting tools into just an old and new version of the tool, or otherwise refactoring them such that common code is in one tool and alternative middle portions are in their own dedicated tool. At the least, putting the contents of those if/else statements into functions would probably help with readability.

marc1uk commented:
as an FYI we're aware of the Travis checks currently failing - it's down to a conflict in preprocessor defines introduced by a ToolDAQFramework change conflicting with other preprocessor defines introduced by boost headers used in ToolAnalysis. For now, using a ToolDAQFramework commit of `a06f13d09845c4f0fb679946f3c385dae406e2fe` or older builds fine.

mnieslony committed:
Added some fixes for the PR

mnieslony commented:
Hey Marcus, Thanks for looking through it.

Going through your comments,
> this seems redundant (getting a variable then setting it to a constant).

I originally thought it might be useful to force matching also at the end of every part file, but decided it was not a good idea. I changed the section now in such a way that the variable `file_completed` is now still obtained from the `CStore` but not used in the decision making for stopping the toolchain.

> if these aren't needed, can we remove them to clean up the code?

Yes, I tried removing most instances of large sections with commented code.

> this seems handy, might be worth putting in ANNIEAlgorithms or something.

Yeah, that could be done. Maybe in a separate PR where one looks through the whole `ANNIEAlgorithms` file

> whats the justification and implications of the 'MRD always being faster'?

The justification is basically that you only have one entry in the MRD sub-BoostStore per event, but you will usually have multiple entries per event for the VME system. So whenever you load the same number of entries for the MRD and VME, the MRD will usually have progressed to a later timestamp already. The implication of this is just that if the event building is in doubt which timestream is lagging behind in the beginning, it will more likely choose the PMT one (which will lead to processing more tank waveforms sooner)

> Personally i'm not sure falling back to a default when given an unrecognised option value is a good idea. 

Do we also want to address this in a separate PR going through all tools? I think there might be multiple tools that feature such a behavior.

> if this is part of the usual event building toolchain it'd be better to replace std::cout with Log with a lower verbosity for frequent printouts. Otherwise a really nicely written tool.

Changed couts to Logs.

> is this something we need to commit? Is this a default toolchain format, or just a debug setting?

Sorry, this was just a remnant from a debugging session. Reverted the change.

> Personally i'm not a fan of toolchains cross-linking their config files.

Yeah, the idea of having default config files in a dedicated directory for the most important configuration files is a good one. Currently I usually just try to cross-link to another toolchain to avoid having to update all configuration files in many different toolchains once something in a common tool changes (especially one as generic/common as the LoadGeometry tool). But having the standard tool configurations for the most important tools in a common directory is a good idea for a future PR.

> Do we need separate toolchains for decoding TandAndMRD, TankAndMRDAndCTC, MRDAndCTC, TankAndCTC, TankAndCTCButNotMRDExceptThoseEventsWithAhhhhSoManyToolchains...?

There are some periods of data for which there was no MRD data for example, so having a separate toolchain that combines Tank and CTC data makes sense for that purpose. Combining MRD and CTC data is also nice to have if you want really small output files and are only interested in the MRD information (for some calibration studies for example). TankAndMRDAndCTC would be the default event building for beam files where you want all the information to be included in the processed data files. But this is of course just a preference and can be discussed.

> dummy files can be removed

Yes, didn't see that one. It should be removed now.

> Please turn on whitespace visibility. 🙃 There's a few others in this PR too (although I leave it to you to go through the 111 files to find them. ;) )

 I tried to remove all the whitespace issues that I was able to spot in the diff files.
 
> Seems like the 'MCFlag' is a bit redundant with the 'is_mc' flag. I would personally prefer some automatic detection, which seems like it ought to be easy enough?

True. Removed `IsMC` from the configuration variables. The MC variable is now set based on whether there is an `MCFlag` entry in the ANNIEEvent store.

> It might be worth considering either splitting tools into just an old and new version of the tool, or otherwise refactoring them such that common code is in one tool and alternative middle portions are in their own dedicated tool. At the least, putting the contents of those if/else statements into functions would probably help with readability.

We could put the if/else blocks in function statements for now and think about possible further improvements. This should probably go in a separate PR though and include also other tools that feature multiple modes of operation.

mnieslony committed:
Added VMEOffset in PMTDataDecoder

mnieslony committed:
some config file changes
